### PR TITLE
Allow Ray API to be used from multiple threads

### PR DESCRIPTION
**Pull request by raulchen**
## What do these changes do?

This PR fixes the issue that Ray API cannot be called from non-main threads by:
1) locking all methods from redis client and plasma client.
2) locking worker's stateful operations, which includes object reconstruction and task index incrementing.

Perf impact for single-thread is neutral:
Benchmark: https://gist.github.com/raulchen/2c315cbd4d2387bf356adbe2a8641986#file-gistfile1-txt
```
With this PR:
test_put_and_get: Avg: 244.46, Min: 206.05, Max: 526.10, P50: 229.27, P90: 267.34
test_remote: Avg: 112.14, Min: 40.58, Max: 257.71, P50: 111.54, P90: 151.92

Without this RP:
test_put_and_get: Avg: 253.18, Min: 214.94, Max: 640.85, P50: 239.45, P90: 276.67
test_remote: Avg: 123.91, Min: 46.94, Max: 328.66, P50: 118.10, P90: 175.88
``` 

## Related issue number
#2385

Previous PR: #2248

### CODE REVIEW

**Comment by pcmoritz** [comment]
can you resolve the merging conflicts in worker.py please?

File: python/ray/worker.py [hunk 1]
```diff
@@ -466,51 +475,56 @@ def get_object(self, object_ids):
...
+                    for unready_id in unready_ids:
+                        self.local_scheduler_client.reconstruct_objects(
+                            [ray.ObjectID(unready_id)], False)
+                    # Do another fetch for objects that aren't available
+                    # locally yet, in case they were evicted since the last
+                    # fetch. We divide the fetch into smaller fetches so as
+                    # to not block the manager for a prolonged period of time
+                    # in a single call.
+                    object_ids_to_fetch = [
+                        plasma.ObjectID(i) for i in unready_ids.keys()
```
Comment (robertnishihara) [inline python/ray/worker.py]: `i` -> `unready_id` or something like that

File: python/ray/worker.py [hunk 2]
```diff
@@ -466,51 +475,56 @@ def get_object(self, object_ids):
...
+                    ]
+                    ray_object_ids_to_fetch = [
+                        ray.ObjectID(i) for i in unready_ids.keys()
```
Comment (robertnishihara) [inline python/ray/worker.py]: `i` -> `unready_id` or something like that

**Comment by robertnishihara** [comment]
@raulchen this looks pretty good. I left some very small comments.

Can you explain where the performance improvement comes from?

**Comment by raulchen** [comment]
Conflict fixed and comment addressed.

@robertnishihara I think the improvement is just random fluctuation. I tried a couple of more times locally, sometimes this PR performs a little better, and sometimes the other way around.

**Comment by robertnishihara** [comment]
jenkins, retest this please

**Comment by robertnishihara** [comment]
The error building the linux wheels looks unrelated. Seems like it failed to pip install cython.

```
[33m  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7feef4e805f8>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/cython/[0m
[33m  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7feef4e80240>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/cython/[0m
[33m  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7feef4e80358>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/cython/[0m
[33m  Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7feef4e804e0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/cython/[0m
[33m  Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7feef4e802b0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /simple/cython/[0m
[31m  Could not find a version that satisfies the requirement cython==0.27.3 (from versions: )[0m
[31mNo matching distribution found for cython==0.27.3[0m
Download error on https://pypi.org/simple/cython/: [Errno -3] Temporary failure in name resolution -- Some packages may not be found!
Download error on https://pypi.org/simple/: [Errno -3] Temporary failure in name resolution -- Some packages may not be found!
No local packages or working download links found for cython<0.28,>=0.27
Traceback (most recent call last):
  File "setup.py", line 156, in <module>
    license="Apache 2.0")
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/setuptools/__init__.py", line 130, in setup
    _install_setup_requires(attrs)
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/setuptools/__init__.py", line 125, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/setuptools/dist.py", line 514, in fetch_build_eggs
    replace_conflicting=True,
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/pkg_resources/__init__.py", line 773, in resolve
    replace_conflicting=replace_conflicting
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1056, in best_match
    return self.obtain(req, installer)
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1068, in obtain
    return installer(requirement)
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/setuptools/dist.py", line 581, in fetch_build_egg
    return cmd.easy_install(req)
  File "/opt/python/cp36-cp36m/lib/python3.6/site-packages/setuptools/command/easy_install.py", line 670, in easy_install
    raise DistutilsError(msg)
distutils.errors.DistutilsError: Could not find suitable distribution for Requirement.parse('cython<0.28,>=0.27')
mv: cannot stat `dist/*.whl': No such file or directory
/ray
/ray/.whl /ray
/ray
```

**Comment by robertnishihara** [comment]
@richardliaw I've seen this failure a couple times now. Any idea about this?

```
+ docker run --rm --shm-size=10G --memory=10G 7229e4d7bcded2b006908682f840bd31bb046567d6e81f8decec1434e321fcbd python /ray/python/ray/rllib/train.py --env Pendulum-v0 --run APEX_DDPG --ray-num-cpus 8
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
Process STDOUT and STDERR is being redirected to /tmp/raylogs/.
Waiting for redis server at 127.0.0.1:40286 to respond...
Waiting for redis server at 127.0.0.1:61642 to respond...
Warning: Reducing object store memory because /dev/shm has only 10737418240 bytes available. You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you may need to pass an argument with the flag '--shm-size' to 'docker run'.
Starting local scheduler with the following resources: {'GPU': 0, 'CPU': 8}.

======================================================================
View the web UI at http://localhost:8888/notebooks/ray_ui714.ipynb?token=2b09154285c759cfef99bb84484502b624b09909f4d33d01
======================================================================

== Status ==
Using FIFO scheduling algorithm.
Result logdir: /root/ray_results/default
PENDING trials:
 - APEX_DDPG_Pendulum-v0_0:	PENDING

Traceback (most recent call last):
  File "/ray/python/ray/rllib/train.py", line 118, in <module>
    run(args, parser)
  File "/ray/python/ray/rllib/train.py", line 112, in run
    queue_trials=args.queue_trials)
  File "/ray/python/ray/tune/tune.py", line 91, in run_experiments
    runner.step()
  File "/ray/python/ray/tune/trial_runner.py", line 118, in step
    trial.config)))
ray.tune.error.TuneError: Insufficient cluster resources to launch trial: trial requested 37 CPUs, 0 GPUs but the cluster only has 8 CPUs, 0 GPUs available. Pass `queue_trials=True` in ray.tune.run_experiments() or on the command line to queue trials until the cluster scales up. 

You can adjust the resource requests of RLlib agents by setting `num_workers` and other configs. See the DEFAULT_CONFIG defined by each agent for more info.

The config of this agent is: {"env": "Pendulum-v0"}
/ray/src/local_scheduler/local_scheduler.cc:177: Killed worker pid 28 which hadn't started yet.
Build step 'Execute shell' marked build as failure
Setting commit status on GitHub for https://github.com/ray-project/ray/commit/cff02ef2adc6b39198374792dc1d658bddb428e6
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6732/
Test FAILed.
Finished: FAILURE
```

**Comment by robertnishihara** [comment]
jenkins, retest this please

**Comment by robertnishihara** [comment]
Seems like there is a test failure in 

```
python test/runtest.py GlobalStateAPI.testFlushAPI
```

I'm not sure what the issue is.

**Comment by robertnishihara** [comment]
Ok, I think the error is likely unrelated. I'll merge this and if the error persists I'll look into it more.

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**